### PR TITLE
added scope visitor

### PR DESCRIPTION
**Pull request by dkrutskikh**
### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix
[ ] New rule
[ ] Changes an existing rule
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

### CODE REVIEW

File: lib/src/models/scoped_class_declaration.dart [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:meta/meta.dart';
+
+import 'class_type.dart';
+
+/// Represents a declaration of a class / mixin / extension.
+@immutable
+class ScopedClassDeclaration {
+  /// Type of declared class entity.
```
Comment (incendial) [inline lib/src/models/scoped_class_declaration.dart:9]: Correction: `The type of the declared class entity`

File: lib/src/models/scoped_class_declaration.dart [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  final ClassType type;
+
+  /// The node that represents in the AST structure for a Dart program.
```
Comment (incendial) [inline lib/src/models/scoped_class_declaration.dart:12]: Seems that there is something missing after `represents` or I didn't catch the meaning...

File: lib/src/models/scoped_class_declaration.dart [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  final CompilationUnitMember declaration;
+
+  /// Returns user defined name.
```
Comment (incendial) [inline lib/src/models/scoped_class_declaration.dart:15]: Correction: `Returns the user defined name`

File: lib/src/models/scoped_class_declaration.dart [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  String get name {
+    if (declaration is ExtensionDeclaration) {
```
Comment (incendial) [inline lib/src/models/scoped_class_declaration.dart:17]: Seems that if you introduce a local variable like `final declaration = this.declaration;` and then type check it, the analyzer will not require a type cast.

File: lib/src/models/scoped_function_declaration.dart [hunk 5]
```diff
@@ -0,0 +1,51 @@
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:meta/meta.dart';
+
+import 'function_type.dart';
+import 'scoped_class_declaration.dart';
+
+/// Represents a declaration of function / method.
+@immutable
+class ScopedFunctionDeclaration {
+  /// Type of declared function entity.
```
Comment (incendial) [inline lib/src/models/scoped_function_declaration.dart:10]: Correction: `The type of the declared function entity.`

File: lib/src/models/scoped_function_declaration.dart [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    if (declaration is FunctionDeclaration) {
+      return (declaration as FunctionDeclaration).name?.name;
+    } else if (declaration is ConstructorDeclaration) {
+      return (declaration as ConstructorDeclaration).name?.name ??
+          (declaration.parent as NamedCompilationUnitMember).name?.name;
+    } else if (declaration is MethodDeclaration) {
+      return (declaration as MethodDeclaration).name?.name;
+    }
+
+    return null;
```
Comment (incendial) [inline lib/src/models/scoped_function_declaration.dart:30]: Why return null here and empty string [here](https://github.com/dart-code-checker/dart-code-metrics/pull/230/files#diff-d7f66dd0ebfb0a270e96a557740e36f9909320394ff085a6520d10cff38b1d97R23)?

File: lib/src/scope_visitor.dart [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  /// Returns the declarations of the found classes.
+  Iterable<ScopedClassDeclaration> get classes => _classes;
+
+  /// Returns the declarations of the found functions.
+  Iterable<ScopedFunctionDeclaration> get functions => _functions;
+
+  @override
+  void visitClassDeclaration(ClassDeclaration node) {
+    _registerClassDeclaration(ClassType.generic, node, () {
+      super.visitClassDeclaration(node);
```
Comment (incendial) [inline lib/src/scope_visitor.dart:25]: Not clear, why do we need to call the base class visit method as a callback?

File: test/resources/class_with_factory_constructors.dart [hunk 8]
```diff
@@ -0,0 +1,22 @@
+class Logger {
+  final String name;
+  bool mute = false;
+
+  // _cache is library-private, thanks to
+  // the _ in front of its name.
```
Comment (incendial) [inline test/resources/class_with_factory_constructors.dart:6]: Do we need this comment in tests?

File: test/scope_visitor_test.dart [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+      final function = visitor.functions.single;
+      expect(function.type, equals(FunctionType.method));
+      expect(function.name, equals('doSomething'));
+      expect(function.fullName, equals('Doer.doSomething'));
+    });
+
+    test('class with factory constructors', () async {
+      (await resolveFile(
+        path: p.normalize(p
+            .absolute('./test/resources/class_with_factory_constructors.dart')),
```
Comment (incendial) [inline test/scope_visitor_test.dart:39]: Suggestion: what do you think about moving path declaration to a separate variable (same for all test cases)?

File: lib/src/scope_visitor.dart [hunk 10]
```diff
@@ -0,0 +1,88 @@
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import 'models/class_type.dart';
+import 'models/function_type.dart';
+import 'models/scoped_class_declaration.dart';
+import 'models/scoped_function_declaration.dart';
+
+/// Visitor to collect declarations of classes and functions.
```
Comment (incendial) [inline lib/src/scope_visitor.dart:9]: Correction: `A visitor to collect declarations of classes and functions.`

File: lib/src/models/scoped_function_declaration.dart [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  /// Returns full user defined name.
```
Comment (incendial) [inline lib/src/models/scoped_function_declaration.dart:33]: Correction: `Returns the full user defined name.`

File: lib/src/models/scoped_function_declaration.dart [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  /// Type of declared function entity.
+  final FunctionType type;
+
+  /// The node that represents in the AST structure for a Dart program.
+  final Declaration declaration;
+
+  /// The class declaration of the class to which this function belongs.
+  final ScopedClassDeclaration enclosingDeclaration;
+
+  /// Returns user defined name.
```
Comment (incendial) [inline lib/src/models/scoped_function_declaration.dart:19]: Correction: `Returns the user defined name.`

**Review by incendial — CHANGES_REQUESTED** [review]

File: test/resources/class_with_factory_constructors.dart (hunk 8 above)
Comment (dkrutskikh) [reply test/resources/class_with_factory_constructors.dart:6]: Yes. This is good example class for future testing some other cases.

**Review by incendial — APPROVED** [review]

**Pull request merged**